src/lib.rs

```diff
@@ -0,0 +1,379 @@
+        // and dereferencing it
+        let ulimit = self.alloc_list.start as usize + SIZE_ALLOC_TABLE;
+        let ptr = self.alloc_list.start as usize + (self.index * core::mem::size_of::<Block>());
+
+        self.index += 1;
+
+        if ptr >= ulimit {
+            return None;
+        }
+
+        let entry = unsafe { &*(ptr as *const Block) };
+        Some(&entry)
+    }
+}
+
+impl AllocList {
+    fn new() -> AllocList {
+        let raw = unsafe { libc::malloc(SIZE_ALLOC_TABLE as libc::size_t) } as *mut Block;
+        AllocList {
+            start: raw,
+            next_free: 0,
+            can_bump: true,
+        }
+    }
+
+    fn iter(&self) -> AllocListIter {
+        AllocListIter {
+            alloc_list: self,
+            index: 0,
+        }
+    }
+
+    fn iter_mut(&mut self) -> AllocListIterMut {
+        AllocListIterMut {
+            alloc_list: self,
+            index: 0,
+        }
+    }
+
+    /// Performs fast bump pointer insertion until the list is full, at which
+    /// point, insertion is O(n) while it linearly scans the list for the next
+    /// free entry.
+    fn insert(&mut self, ptr: usize, size: usize) {
+        if self.can_bump {
+            unsafe {
+                let next_ptr = self.start.add(self.next_free) as *mut Block;
+                if (next_ptr as usize) < self.start as usize + SIZE_ALLOC_TABLE {
+                    *next_ptr = Block::Entry(core::num::NonZeroUsize::new_unchecked(ptr), size);
+                    self.next_free += 1;
+                    return;
+                } else {
+                    self.can_bump = false;
+                }
+            }
+        }
+
+        // Slow path, we need to linearly scan for the next free block in the
+        // heap.
+        for block in self.iter_mut() {
+            match block {
+                Block::Free => {
+                    *block =
+                        Block::Entry(unsafe { core::num::NonZeroUsize::new_unchecked(ptr) }, size);
+                    return;
+                }
+                _ => (),
+            }
+        }
+
+        // The allocation list is full
+        abort();
+    }
+
+    /// Remove ptr information associated with a base pointer (perfomed on a
+    /// dealloc). This must not be called with an inner pointer.
+    fn remove(&mut self, ptr: usize) {
+        for block in self.iter_mut() {
+            match *block {
+                Block::Entry(base_ptr, _) => {
+                    if base_ptr.get() == ptr {
+                        *block = Block::Free;
+                        return;
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+
+    /// Given an arbitrary pointer (base or inner), finds pointer info of the
+    /// associated base pointer if it exists.
+    fn find_base(&self, ptr: usize) -> Option<PtrInfo> {
+        for block in self.iter() {
```

Comment:
Might it be better to use `self.iter().position(...)` as I think the compiler can generally better optimise that than a`for` loop?

Reply:
Ah, maybe then just `find` or equivalent. Basically the compiler has a better chance, in general, of optimising those calls then it does a generic `for` loop.

Reply:
The new approach uses `find_map`.